### ISSUE TITLE
Seg faulting on index that shouldn't seg fault

### ISSUE THREAD

**Issue by JoshWolper** [taichi-dev/taichi#2146]
**Describe the bug**
Getting a seg fault on my data structure when using an in-range index. This is an attempt to implement a 2D/3D spatial hash following a very similar approach to the one here: https://github.com/taichi-dev/taichi/blob/master/examples/pbf2d.py
Also worth noting that this bug does not occur for 2D, only 3D!

**To Reproduce**

```py
import taichi as ti
import numpy as np
import triangle as tr

#ti.init(default_fp=ti.f64, arch=ti.gpu) # Try to run on GPU    #GPU, parallel
ti.init(default_fp=ti.f64, arch=ti.cpu, cpu_max_num_threads=1)  #CPU, sequential

#Sparse Grids
#---Params
dim = 3
dx = 0.003
rp = (3*(dx**2))**0.5
invDx = 1.0 / 0.003
nGrid = ti.ceil(invDx)
grid_size = 4096
grid_block_size = 128
leaf_block_size = 16 if dim == 2 else 8
indices = ti.ij if dim == 2 else ti.ijk
offset = tuple(0 for _ in range(dim))
#---Grid Shapes for PID
grid = ti.root.pointer(indices, grid_size // grid_block_size) # 32
block = grid.pointer(indices, grid_block_size // leaf_block_size) # 8
pid = ti.field(int)
block.dynamic(ti.indices(dim), 1024 * 1024, chunk_size=leaf_block_size**dim * 8).place(pid, offset=offset + (0, ))
#---Grid Shapes for Rest of Grid Structures
grid2 = ti.root.pointer(indices, grid_size // grid_block_size) # 32
block2 = grid2.pointer(indices, grid_block_size // leaf_block_size) # 8
def block_component(c):
    block2.dense(indices, leaf_block_size).place(c, offset=offset) # 16 in 3D, 8 in 2D (-2048, 2048) or (0, 4096) w/o offset

#Grid Structures
gridNumParticles = ti.field(dtype=int)      #track number of particles in each cell using cell index
maxPPC = 2**10
block_component(gridNumParticles) #keep track of how many particles are at each cell of backGrid
backGrid = ti.field(int)              #background grid to map grid cells to a list of particles they contain
backGridIndeces = ti.ijk if dim == 2 else ti.ijkl
backGridShape = (nGrid, nGrid, maxPPC) if dim == 2 else (nGrid, nGrid, nGrid, maxPPC)
ti.root.dense(backGridIndeces, backGridShape).place(backGrid)      #backGrid is nGrid x nGrid x maxPPC

#Particle Structures
x = ti.Vector.field(dim, dtype=float) # position
mp = ti.field(dtype=float) # particle masses
particle = ti.root.dynamic(ti.i, 2**27, 2**19) #2**20 causes problems in CUDA (maybe asking for too much space)
particle.place(x, mp)

#Neighbor Search Routines
@ti.func
def backGridIdx(x):
    #compute int vector of backGrid indeces (recall the grid here )
    return int(x/rp)

@ti.func
def stencil_range():
    return ti.ndrange(*((3, ) * dim))

def addParticles():
    N = 10
    w = 0.2
    dw = w / N
    for i in range(N):
        for j in range(N):
            for k in range(N):
                x[i*(N**2) + j*N + k] = [0.4 + (i * dw), 0.4 + (j * dw), 0.4 + (k * dw)]
                mp[i*(N**2) + j*N + k] = 0.001
            
@ti.kernel
def build_pid():
    ti.block_dim(64) #this sets the number of threads per block / block dimension
    for p in x:
        base = int(ti.floor(x[p] * invDx - 0.5))
        ti.append(pid.parent(), base - ti.Vector(list(offset)), p)

@ti.kernel
def backGridSort():
    #Sort particles into backGrid
    ti.block_dim(256)
    ti.no_activate(particle)
    for I in ti.grouped(pid):
        p = pid[I]
        cell = backGridIdx(x[p]) #grab cell idx (vector of ints)
        offs = ti.atomic_add(gridNumParticles[cell], 1) #atomically add one to our grid cell's particle count NOTE: returns the OLD value before add
        print("cell:", cell, "offs:", offs)
        print("backGrid shape:", backGrid.shape)
        backGrid[cell, offs] = p #place particle idx into the grid cell bucket at the correct place in the cell's neighbor list (using offs)
 
addParticles()
grid.deactivate_all()
build_pid()
backGridSort()

```

**Log/Screenshots**
Please post the **full log** of the program (instead of just a few lines around the error message, unless the log is > 1000 lines). This will help us diagnose what's happening. For example:

```
[Taichi] mode=release
[Taichi] preparing sandbox at /tmp/taichi-86oajjkm
[Taichi] version 0.6.37, llvm 10.0.0, commit e0ac1d86, linux, python 3.6.9
[Taichi] Starting on arch=x64
[Taichi] materializing...
cell: [76, 76, 76] offs: 0
backGrid shape: [334, 334, 334, 1024]
[E 01/08/21 20:34:28.869] Received signal 11 (Segmentation fault)


***********************************
* Taichi Compiler Stack Traceback *
***********************************
/home/josh/.local/lib/python3.6/site-packages/taichi/core/../lib/taichi_core.so: taichi::Logger::error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)
/home/josh/.local/lib/python3.6/site-packages/taichi/core/../lib/taichi_core.so: taichi::signal_handler(int)
/lib/x86_64-linux-gnu/libc.so.6(+0x3efd0) [0x7fd521d29fd0]
[0x7fd522198f36]

Internal Error occurred, check this page for possible solutions:
https://taichi.readthedocs.io/en/stable/install.html#troubleshooting

...
```

**Comment by k-ye** [comment]
Hi,

Could you provide a bit more info? For example:

1. does this occur only when the grid size is very large (`4096`)?
1. If this still happens on a small scale, what if you replace all the `pointer` to `dense`?

**Comment by yuanming-hu** [comment]
A few other notes: 
- If you are running on CUDA, try turning off unified memory and preallocate a large chunck of memory: `ti.init(arch=ti.cuda, use_unified_memory=False, device_memory_fraction=0.8)`. This may help you get rid of the "maybe asking for too much space" issue
- Use `ti.init(debug=True)` to enable bound checks that help diagnose access out of bounds, which can be a common reason for segmentation faults.

**Comment by JoshWolper** [comment]
Hi again y'all, thanks for the suggestions! I checked some other smaller grid sizes and it appears the bug is independent of the grid size. Also I'm running this on CPU with just one thread for now, so no CUDA! I also did try changing gridNumParticles to be densely placed similarly to backGrid, but this has the same bug. I also ran after adding debug=True to my init parameters and it did not appear to add anything new to the errors :( Based on your comment I thought that maybe the problem was the disparity in structure between gridNumParticles and backGrid but I guess not! I just can't seem to figure out why the 3D version of this doesn't work but 2d does, could it be related to ti.ijkl somehow?

**Comment by k-ye** [comment]
> `version 0.6.37`: 

Try upgrading `taichi` to the latest? (v0.7.12)

**Comment by k-ye** [comment]
Ah, `backGrid` is a `dense` with shape `[334, 334, 334, 1024]`, which takes `334x334x334x1024xsizeof(int32) ~= 152GB`... I think Taichi also pads any non-power-of-two dimension, so it's `512^3*1024*4`... :fearful:  Maybe try make `backGrid` a `ti.root.pointer((334,334,334)).dense(1024)`?

That said, I'm trying to figure out why this OOM wasn't reported when materializing the fields in Taichi. That way such kind of errors should be much easier to identify. (Also my claim that this was not reproduced in https://github.com/taichi-dev/taichi/issues/2146#issuecomment-758558516 was done on Windows. Not sure if and how virtual memory/lazy allocation is playing here..)

**Comment by JoshWolper** [comment]
Oh wow okay well this explains the difficult to explain behavior!! I actually was trying to set it up the way you say like `ti.root.pointer((334,334,334)).dense(1024)`, but I don't how to do this because this would require ti.L wouldn't it? If I try to do it like this: `ti.root.pointer(backGridIndeces, nGrid).dense(???, maxPPC).place(backGrid)` what do I put for the indeces of dense? The reason I allocate it the previous way was that it allowed me to directly use ti.ijkl

**Comment by k-ye** [comment]
Awesome, FYI here are all the predefined indices:

https://github.com/taichi-dev/taichi/blob/ddcf000170f5c6247ead496a176ae9de3dc39744/python/taichi/lang/__init__.py#L15-L26

---

I've found the root cause of the problem. It's still because of the `int32` -- The 4-D coordinate is linearized into an `i32`, which went negative. Here's a small repro:

```py
import taichi as ti
import numpy as np

#ti.init(default_fp=ti.f64, arch=ti.gpu) # Try to run on GPU    #GPU, parallel
ti.init(default_fp=ti.f32, arch=ti.cpu, cpu_max_num_threads=1, print_ir=True)  #CPU, sequential

dim = 3
invDx = 1.0 / 0.003
nGrid = ti.ceil(invDx)

maxPPC = 2**10

backGridIndeces = ti.ijk if dim == 2 else ti.ijkl
backGridShape = (nGrid, nGrid, maxPPC) if dim == 2 else (nGrid, nGrid, nGrid, maxPPC)
backGrid = ti.field(int, backGridShape)
# ti.root.dense(backGridIndeces, backGridShape).place(backGrid)      #backGrid is nGrid x nGrid x maxPP

@ti.kernel
def printInteresting():
  print(backGrid[27, 301, 301, 0])  # boom
  # print(backGrid[1,2,3,4])  # this is fine

printInteresting()
```

Corresponding Taichi IR:

```
kernel {
  $0 = offloaded  
  body {
    <i32> $1 = const [0]
    <*gen> $2 = get root
    <*gen> $3 = [S0root][root]::lookup($2, $1) activate = false
    <*gen> $4 = get child [S0root->S1dense] $3
    <i32> $5 = const [-1184058368]  <----- !!
    <*gen> $6 = [S1dense][dense]::lookup($4, $5) activate = false
    <*i32> $7 = get child [S1dense->S2place<i32>] $6
    <i32> $8 = global load $7
    print $8, "\n"
  }
}
```

Changing this to `i64` requires some amount of efforts. Instead, I'll try to add some boundary checks when `debug=True` is set.

**Comment by k-ye** [comment]
This has been (partially) fixed in #2199  by @leon5288. If you turn on `debug=True`, it will dump a huge warning now.
